### ISSUE TITLE
KNN affinity model

### ISSUE THREAD

**Issue by dkobak** [pavlin-policar/openTSNE#131]
I am often playing with kNN affinity kernel, which I emulate like this: `affinity.FixedSigmaNN(X, sigma=1e+6, k=15)`. How about making a separate class `affinity.KNN()` that would assign equal p_i|j values to all kNN neighbors, and then normalize and symmetrize as usual?

**Comment by pavlin-policar** [comment]
The API is written so that this kind of stuff is fairly simple. You can define a class like this (I haven't tested this at all, but it should be something like this) and pass this as a parameter to the `TSNEEmbedding` constructor. 

```python
class KNNGAffinities(openTSNE.affinity.Affinities):
    def __init__(
        self,
        data,
        k_neighbors=30,
        method="auto",
        metric="euclidean",
        metric_params=None,
        symmetrize=True,
        n_jobs=1,
        random_state=None,
        verbose=False,
    ):
        self.n_samples = data.shape[0]
        self.k_neighbors = k_neighbors
        self.verbose = verbose
        self.n_jobs = n_jobs

        self.knn_index, neighbors, distances = openTSNE.affinity.build_knn_index(
            data, method, k_neighbors, metric, metric_params, n_jobs, random_state, verbose
        )
        A = sp.csr_matrix(
            (
                np.ones_like(distances).ravel(),
                neighbors.ravel(),
                range(0, self.n_samples * self.k_neighbors + 1, self.k_neighbors),
            ),
            shape=(self.n_samples, self.n_samples),
        )

        # Symmetrize + normalize the probability matrix
        P = (A + A.T) / 2
        P /= np.sum(P)
        self.P = P
```

I am a bit hesitant about just adding a bunch of affinity models because they haven't really been shown to work. I had also played around with an affinity model that took any undirected graph, which is also interesting. Perhaps it might be better to start a wiki page here on the GH repository and add code snippets there.

**Comment by dkobak** [comment]
For me this one is important because for high-dimensional data using kNN with k=90 becomes a major bottleneck. T-SNE usually produces almost the same result with perplexity 30 (meaning k=90) and uniform kNN kernel with k=10--15 (some evidence here: https://twitter.com/hippopedoid/status/1232698023253303298). In most cases I do not care much about it because I reduce the dimensionality with PCA and then it's not expensive to find k=90 neighbors. But now when playing with sparse data and cosine distances, I am not using PCA first, and so end up needing to find kNN with Pynndescent directly on a pretty large matrix. Finding k=90 neighbors makes it ridiculously long, compared to k=15 or k=10 neighbors. And of course UMAP ends up much faster because it only uses k=15 neighbors.

**Comment by dkobak** [comment]
An alternative idea that came to my mind today is to add a `n_neighbors` parameter that would explicitly define the k in the kNN graph, separate from perplexity. The default setting would be n_neighbors = 3 * perplexity. But this is of course not set in stone, and one could use 2*perplexity or 10*perplexity if one wants. In particular, one could use n_neighbors = perplexity, in which case the sigma becomes infinite and the distribution uniform, i.e. it's the kNN kernel. So this would allow to specify a kNN affinity kernel using e.g. `perplexity=15, n_neighbors=15`. Or one could use something like `perplexity=10, n_neighbors=15` which will be slightly decaying but still restricted to 15 nearest neighbors.

**Comment by pavlin-policar** [comment]
I really don't want to put this into the standard, public API. This is quite different from what we currently do, and I don't really see the benefits of this for the average user, who just wants a pretty picture. I also don't think this would have a very drastic effect on the embeddings (you said once that the uniform kernel produces very similar results to the standard gaussian, and this seems to be a kind of middle ground). So I definitely don't want to add that into the public API.

That said, it seems like an interesting thing to try out, and this is exactly the kind of stuff openTSNE was designed to make easy to test. It should be easy to put together an affinity model that would do this. And if this ends up being useful, we can put it into `openTSNE.affinity`. This way people would still be able to access it, but they would know what they're getting into.

I don't really see what the benefit of this is though. Why would setting `k_neighbors=10*perplexity` be a good idea? Since these are put into the Gaussian, these decay exponentially, and anything over `3*perplexity` would be virtually zero. Or would you also change how the kernel bandwidth is calculated? In that case, perplexity wouldn't make that much sense anymore. As for the uniform kernel, I like this, but it feels like it should be its own affinity model. I also like the idea of being able to pass in any graph since t-SNE is just a force-directed layout and compare that to graph-layout algorithms. Maybe these could be combined in some way.

**Comment by dkobak** [comment]
My main point was that it can take a lot more time to build kNN graph with k=90 compared to k=15 for large sparse matrices. E.g. analyzing a 100000x9630 sparse matrix with prime factorizations on an old laptop now, I got the following runtimes (after numba precompilation):
```
UMAP(metric='cosine', init='random').fit_transform(X)    # 8 min
TSNE(metric='cosine', n_jobs=-1, initialization='random').fit(X)   # 24 min
```
so deafult t-SNE is a lot slower. How to make it faster? Use smaller k. My go-to approach is to use uniform kernel with k=15 (or even 10), like this:
```
A = affinity.FixedSigmaNN(X, sigma=1e+6, k=15, metric="cosine", n_jobs=-1)
I = initialization.random(X)
E = TSNEEmbedding(I, A, n_jobs=-1)
E = E.optimize(n_iter=250, exaggeration=12, momentum=0.5)
E = E.optimize(n_iter=500, exaggeration=1,  momentum=0.8)
```
(4.5 min), but I think it's pretty cumbersome for any normal user. Note that it will not become much less cumbersome even if uniform affinity kernel gets its own class. My suggestion to include `n_neighbors` parameter would allow a one-liner `fit()` call with 15 neighbors and either a uniform affinity kernel (perplexity=15) or a slightly decaying kernel (perplexity=10). Based on my experiments, the result should be very similar to the default perplexity=30.

Of course there is an alternative approach, which is to use perplexity=5. It would use k=15:
```
TSNE(metric='cosine', perplexity=5, n_jobs=-1, initialization='random').fit(X)
```
(4.5 min). Maybe this is fine, but I am usually a bit reluctant to use so small perplexity values.

This was just to explain my reasoning here.

**Comment by pavlin-policar** [comment]
Oh, I see, that makes sense then. I didn't think you were coming at it from an execution-time perspective.

Yeah, the long way of writing it out is a bit annoying, I notice it when I want to do multiscale kernels as well. Then having to call optimize every time can get tedious. I still think adding a `k_neighors` parameter to the default implementation is a bad idea. We could make this a whole lot less annoying by adding an `affinities` parameter to the constructor and then the 5-line t-SNE call would be reduced to 2 lines. I like this a lot more than adapting the default API to the uniform kernel. What do you think?

**Comment by dkobak** [comment]
Sounds good but I am not sure I fully understand the idea. Can you show how the call would look like then?

**Comment by pavlin-policar** [comment]
Sure thing, a call would then look something like

```python
affinity = openTSNE.affinity.UniformKernel(X, k_neighbors=15, n_jobs=-1)
embedding = openTSNE.TSNE(affinity=affinity, n_jobs=-1).fit(X)
```

This still requires `n_jobs` to be specified twice, but we might be able to figure a way around that. This is a lot nicer than the 5-liner that we usually have to write out and it makes it easy to use any affinity model.

**Comment by dkobak** [comment]
Yes, I like this a lot. This would be a huge improvement to the API because currently one needs to know all these default `optimize()` parameters by heart. Specifying `n_jobs` twice is okay.

Adding uniform kNN affinity model class would also be great.

And we could still add `n_neighbors` to the `PerplexityBasedNN` or maybe something like `n_neighbors_factor` to replace `3` in https://github.com/pavlin-policar/openTSNE/blob/master/openTSNE/affinity.py#L136, but this is then not really necessary, just for completeness.

**Comment by dkobak** [comment]
Should we go ahead with this? Do you want me to prepare a PR?

**Comment by dkobak** [comment]
Great that it's done now. Are you planning to put a 0.4.4 tag?

**Comment by pavlin-policar** [comment]
Yeah, for sure. I'm just working on some figures and am making sure everything is working as intended. I am holding off a release in case any bugs pop up, but I'll definitely release a 0.4.4 in a few days or so.

**Comment by dkobak** [comment]
Great!
